### PR TITLE
Rewrite Rules: Add opt-in support for random content redirects.

### PR DESCRIPTION
**Pull request by mlaetitia**
Trac ticket: https://core.trac.wordpress.org/ticket/64498

This PR adds a minimal, opt-in implementation of **random content redirects**

## What it does

* Adds `wp_random_content_redirect()` (in `wp-includes/query.php`, alongside its closest sibling `wp_old_slug_redirect()`), hooked on `template_redirect`.
* When enabled, a request to `example.com/?random` issues a 302 redirect to a randomly selected published post. `?random&random_post_type=page` restricts the pick to a post type (viewable types only); `?random&random_cat_id=<term ID>` restricts it to a category. These parameter names match Matt's original [Random Redirect plugin](https://wordpress.org/plugins/random-redirect/) referenced in the ticket, so existing `?random` links keep working.
* **Disabled by default** — per the ticket's request for an opt-in implementation with no user-facing settings or defaults, nothing happens unless a theme or plugin opts in:

  ```php
  add_filter( 'enable_random_content_redirect', '__return_true' );
  ```

* A second filter, `random_content_redirect_url`, allows short-circuiting or rewriting the redirect target (same pattern as `old_slug_redirect_url`).

## Implementation notes

* The random pick uses a COUNT plus a random OFFSET (two `WP_Query` calls) rather than `ORDER BY RAND()`, which randomizes and sorts every candidate row on each request and is uncached.
* Password-protected and non-published posts are never redirect targets. Non-viewable post types are rejected. Only GET/HEAD requests are handled.
* The redirect is a 302 so clients do not cache the randomly picked target.
* Open question for review: the parameters are read from `$_GET` on `template_redirect`, matching the original plugin's addressing. If pretty-URL support (e.g. a `/random/` rewrite endpoint) is desirable, the parameters should instead become registered query vars — happy to rework in that direction if preferred.

## Tests

`tests/phpunit/tests/rewrite/randomContentRedirect.php` covers: disabled-by-default behavior, no-op without the `random` parameter, redirect to a published post, post type and category restriction, exclusion of password-protected/draft/private posts, invalid and non-viewable post types, nonexistent categories, empty result sets, and non-GET requests.

## Use of AI Tools

AI assistance: Yes
Tool(s): Claude Code
Model(s): Claude Fable 5
Used for: Initial implementation and test suite, ported from the Jetpack Random Redirect module under my direction; reviewed, tested, and edited by me.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/query.php [hunk 1]
```diff
@@ -1235,6 +1235,119 @@ function _find_post_by_old_date( $post_type ) {
...
+	 * @param string $link The redirect URL.
+	 */
+	$link = apply_filters( 'random_content_redirect_url', $link );
+
+	if ( ! $link ) {
+		return;
+	}
+
+	// Temporary redirect, so clients do not cache the randomly picked target.
+	wp_safe_redirect( $link, 302 );
```
Comment (ellatrix) [inline src/wp-includes/query.php:1350]: Should we also call `nocache_headers`?

File: src/wp-includes/query.php [hunk 2]
```diff
@@ -1235,6 +1235,119 @@ function _find_post_by_old_date( $post_type ) {
...
+		'order'                  => 'ASC',
+		'ignore_sticky_posts'    => true,
+		'update_post_term_cache' => false,
+		'update_post_meta_cache' => false,
+	);
+
+	if ( isset( $_GET['random_cat_id'] ) ) {
+		$cat_id = absint( wp_unslash( $_GET['random_cat_id'] ) );
+
+		if ( $cat_id < 1 ) {
```
Comment (ellatrix) [inline src/wp-includes/query.php:1303]: `absint` always returns positive? Should it be `(int)` instead maybe?

File: tests/phpunit/tests/rewrite/randomContentRedirect.php [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+		$_SERVER['REQUEST_METHOD'] = 'POST';
+
+		self::factory()->post->create();
+
+		wp_random_content_redirect();
+		$this->assertNull( $this->random_content_redirect_url );
+	}
+
+	public function test_head_request_redirects() {
+		$_SERVER['REQUEST_METHOD'] = 'HEAD';
```
Comment (ellatrix) [inline tests/phpunit/tests/rewrite/randomContentRedirect.php:80]: Do we need to restore this?

File: tests/phpunit/tests/rewrite/randomContentRedirect.php [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+		wp_random_content_redirect();
+		$this->assertSame( get_permalink( $public_id ), $this->random_content_redirect_url );
+	}
+
+	public function test_random_post_type_parameter() {
+		self::factory()->post->create();
+		$page_id = self::factory()->post->create( array( 'post_type' => 'page' ) );
+
+		$_GET['random_post_type'] = 'page';
```
Comment (ellatrix) [inline tests/phpunit/tests/rewrite/randomContentRedirect.php:115]: Same here, not sure if these are restored automatically.

File: src/wp-includes/query.php [hunk 5]
```diff
@@ -1235,6 +1235,119 @@ function _find_post_by_old_date( $post_type ) {
...
+	 *
+	 * @since 7.2.0
+	 *
+	 * @param bool $enabled Whether `?random` requests redirect to random content. Default false.
+	 */
+	if ( ! apply_filters( 'enable_random_content_redirect', false ) ) {
+		return;
+	}
+
+	if ( isset( $_SERVER['REQUEST_METHOD'] ) && ! in_array( strtoupper( $_SERVER['REQUEST_METHOD'] ), array( 'GET', 'HEAD' ), true ) ) {
```
Comment (ellatrix) [inline src/wp-includes/query.php]: Not sure if by design, but if `$_SERVER['REQUEST_METHOD']` isn't set, it will not return early.

**Review by ellatrix — APPROVED** [review]
Just some nits, leaving some time for others to see this and can commit if you're ready.

File: src/wp-includes/query.php (hunk 5 above)
Comment (mcsf) [reply src/wp-includes/query.php]: When would it not be set?

File: src/wp-includes/query.php (hunk 1 above)
Comment (mcsf) [reply src/wp-includes/query.php:1350]: I suppose the intent here is to not cache the redirection (hence 302), while allowing clients to cache the resolved page (i.e. the randomly selected post). No?

File: tests/phpunit/tests/rewrite/randomContentRedirect.php [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+		$post_id = self::factory()->post->create();
+
+		wp_random_content_redirect();
+		$this->assertSame( get_permalink( $post_id ), $this->random_content_redirect_url );
+	}
+
+	public function filter_random_content_redirect_url( $url ) {
+		$this->random_content_redirect_url = $url;
+		return false;
+	}
```
Comment (mcsf) [inline tests/phpunit/tests/rewrite/randomContentRedirect.php]: Nit: for readability, I recommend moving this method above the `test_*` ones.

File: tests/phpunit/tests/rewrite/randomContentRedirect.php [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+	public function test_disabled_by_default() {
+		remove_filter( 'enable_random_content_redirect', '__return_true' );
+
+		self::factory()->post->create();
+
+		wp_random_content_redirect();
+		$this->assertNull( $this->random_content_redirect_url );
+	}
```
Comment (mcsf) [inline tests/phpunit/tests/rewrite/randomContentRedirect.php:33]: Because of the setup and the presence of `remove_filter` at the beginning of the test, the test isn't actually testing anything meaningful. :)

I know it's annoying, but the most straightforward and verifiable way to write this whole test suite would be to:

- remove `add_filter( 'enable_random_content_redirect', '__return_true' )` from `set_up`
- remove `remove_filter(...)` from this test, so as to test the default conditions
- start each subsequent test with `add_filter`

File: tests/phpunit/tests/rewrite/randomContentRedirect.php (hunk 3 above)
Comment (mcsf) [reply tests/phpunit/tests/rewrite/randomContentRedirect.php:80]: I had the same question about superglobals and hooks, but I then saw that everything is saved and restored:

https://github.com/WordPress/wordpress-develop/blob/2b2b6a918784a67bb19c6e166fea0a683edba474/tests/phpunit/includes/abstract-testcase.php#L119-L123

https://github.com/WordPress/wordpress-develop/blob/2b2b6a918784a67bb19c6e166fea0a683edba474/tests/phpunit/includes/abstract-testcase.php#L230

File: tests/phpunit/tests/rewrite/randomContentRedirect.php (hunk 4 above)
Comment (mcsf) [reply tests/phpunit/tests/rewrite/randomContentRedirect.php:115]: They are; see above.

File: src/wp-includes/query.php (hunk 1 above)
Comment (ellatrix) [reply src/wp-includes/query.php:1350]: Wouldn't it cache based on the URL? I'm not a page cache expert, hence my comment being formatted as a question 🙂

Comment (mlaetitia) [reply src/wp-includes/query.php:1350]: Thanks both for the comments. I ended up adding a `nocache_headers` as we actually want both as much as possible:
- the 302 makes the redirect non-permanent
- stop intermediary caches (CDNs, full-page caches) from pinning the redirect, which would make ?random return the same post for everyone

Thanks! Hadn't considered this properly

File: src/wp-includes/query.php (hunk 2 above)
Comment (mlaetitia) [reply src/wp-includes/query.php:1303]: Great catch! I changed this and also on the same note changed the test that checked the negative value as it was generally failing (as it should) but mostly for the wrong reasons.
changed both on the last commit

File: src/wp-includes/query.php (hunk 5 above)
Comment (mlaetitia) [reply src/wp-includes/query.php]: Generally it wouldn't (maybe some CLI scenarios?) , but I rewrote this to explicitly accept only GET or HEAD.
Note: I didn't include the new QUERY method since Core doesn't yet support it (so can't imagine a scenario where it would be sent)

File: tests/phpunit/tests/rewrite/randomContentRedirect.php (hunk 6 above)
Comment (mlaetitia) [reply tests/phpunit/tests/rewrite/randomContentRedirect.php]: Done

File: tests/phpunit/tests/rewrite/randomContentRedirect.php (hunk 7 above)
Comment (mlaetitia) [reply tests/phpunit/tests/rewrite/randomContentRedirect.php:33]: Yap - I agree ! changed this in the latest commit - while annoying definitely makes more sense this way. thanks for catching that

File: tests/phpunit/tests/rewrite/randomContentRedirect.php (hunk 3 above)
Comment (mcsf) [reply tests/phpunit/tests/rewrite/randomContentRedirect.php:80]: Oh wait, no, `$_SERVER` isn't included here, but my brain really wanted it to be:

https://github.com/WordPress/wordpress-develop/blob/2b2b6a918784a67bb19c6e166fea0a683edba474/tests/phpunit/includes/abstract-testcase.php#L241-L246

**Review by mcsf — CHANGES_REQUESTED** [review]
This looks good to me, but I'd feel extra confident if maybe @desrosj had a look too — if you can spare a moment. :)

File: tests/phpunit/tests/rewrite/randomContentRedirect.php (hunk 3 above)
Comment (mlaetitia) [reply tests/phpunit/tests/rewrite/randomContentRedirect.php:80]: Thanks for the second check @mcsf  - if I am not mistaken `$_SERVER` variables are cleaned later on in (assuming `WP_RUN_CORE_TESTS` is set):

https://github.com/WordPress/wordpress-develop/blob/2b2b6a918784a67bb19c6e166fea0a683edba474/tests/phpunit/includes/functions.php#L24-L35

Would you prefer if I added a `tear_down` anyway?

Comment (mcsf) [reply tests/phpunit/tests/rewrite/randomContentRedirect.php:80]: Thanks, @mlaetitia! OK, that looks good then, might not need explicit tearing down.

(I suggest verifying locally that the global is correctly reset, for good measure, but analytically it looks good.)

**Review by mcsf — APPROVED** [review]

**Comment by annezazu** [comment]
@desrosj nudging here for when you have a chance to review.

File: src/wp-includes/query.php [hunk 8]
```diff
@@ -1235,6 +1235,122 @@ function _find_post_by_old_date( $post_type ) {
...
+	$link = apply_filters( 'random_content_redirect_url', $link );
+
+	if ( ! $link ) {
+		return;
+	}
+
+	// Temporary redirect, so clients and intermediary caches do not cache the randomly picked target.
+	nocache_headers();
+	wp_safe_redirect( $link, 302 );
+	exit;
```
Comment (westonruter) [inline src/wp-includes/query.php:1351]: To facilitate unit test coverage of this line, this could instead be:


```suggestion
	if ( wp_safe_redirect( $link, 302 ) ) {
		exit;
	}
```

File: src/wp-includes/query.php [hunk 9]
```diff
@@ -1235,6 +1235,122 @@ function _find_post_by_old_date( $post_type ) {
...
+ *
+ * Disabled by default. To enable:
+ *
+ *     add_filter( 'enable_random_content_redirect', '__return_true' );
+ *
+ * @since 7.2.0
+ */
+function wp_random_content_redirect() {
+	// The `random` parameter is a flag, present with or without a value.
+	if ( ! isset( $_GET['random'] ) ) {
```
Comment (westonruter) [inline src/wp-includes/query.php:1255]: Instead of reading from `$_GET`, shouldn't `random` be added to `\WP::$public_query_vars` and then here it can do:

```suggestion
	if ( ! get_query_var( 'random' ) ) {
```

File: src/wp-includes/query.php [hunk 10]
```diff
@@ -1235,6 +1235,122 @@ function _find_post_by_old_date( $post_type ) {
...
+
+			$query = new WP_Query( $args );
+		}
+	}
+
+	if ( empty( $query->posts ) ) {
+		return;
+	}
+
+	$link = get_permalink( $query->posts[0] );
```
Comment (westonruter) [inline src/wp-includes/query.php:1331]: Since plugins might filter the `the_posts` filter to remove items from the array, it may no longer be a list and `$query->posts[0]` could be undefined. It would be better to use `array_first()`:

```suggestion
	$link = get_permalink( array_first( $query->posts ) );
```

File: src/wp-includes/query.php [hunk 11]
```diff
@@ -1235,6 +1235,122 @@ function _find_post_by_old_date( $post_type ) {
...
+
+	$request_method = isset( $_SERVER['REQUEST_METHOD'] ) ? strtoupper( $_SERVER['REQUEST_METHOD'] ) : '';
+
+	if ( ! in_array( $request_method, array( 'GET', 'HEAD' ), true ) ) {
+		return;
+	}
+
+	$post_type = 'post';
+
+	if ( isset( $_GET['random_post_type'] ) ) {
```
Comment (westonruter) [inline src/wp-includes/query.php:1278]: Ditto above.

Comment (westonruter) [reply src/wp-includes/query.php:1278]: Also, this should ideally check if `$_GET['random_post_type']` is a string. Someone could make a request to a page like `?random=1&random_post_type[]=error` and then `$_GET['random_post_type']` here would be `array( 'error' )`. This array would get passed into `sanitize_key()` as an array, which would then return an empty string so that avoids a fatal error. However, the `sanitize_key` filter includes a second param with the original key passed, so a callback could fatal if it is expecting a string but an array is passed.

Related: Core-60745.


```suggestion
	if ( isset( $_GET['random_post_type'] ) && is_string( $_GET['random_post_type'] ) ) {
```

File: src/wp-includes/query.php [hunk 12]
```diff
@@ -1235,6 +1235,122 @@ function _find_post_by_old_date( $post_type ) {
...
+		'has_password'           => false,
+		'posts_per_page'         => 1,
+		'orderby'                => 'ID',
+		'order'                  => 'ASC',
+		'ignore_sticky_posts'    => true,
+		'update_post_term_cache' => false,
+		'update_post_meta_cache' => false,
+	);
+
+	if ( isset( $_GET['random_cat_id'] ) ) {
```
Comment (westonruter) [inline src/wp-includes/query.php:1300]: Same as above about using `get_query_var()` but also for type safety, which is less important here because of the `(int)` cast:

```suggestion
	if ( isset( $_GET['random_cat_id'] ) && is_string( $_GET['random_cat_id'] ) ) {
```

File: src/wp-includes/query.php [hunk 13]
```diff
@@ -1235,6 +1235,122 @@ function _find_post_by_old_date( $post_type ) {
...
+	 *
+	 * @since 7.2.0
+	 *
+	 * @param string $link The redirect URL.
+	 */
+	$link = apply_filters( 'random_content_redirect_url', $link );
+
+	if ( ! $link ) {
+		return;
+	}
```
Comment (westonruter) [inline src/wp-includes/query.php:1346]: I wonder if we need this? There is already a filter applying in `wp_safe_redirect()` including `wp_redirect`. The use of this filter for unit testing can be replaced with using `wp_redirect`.

File: src/wp-includes/query.php [hunk 14]
```diff
@@ -1235,6 +1235,122 @@ function _find_post_by_old_date( $post_type ) {
...
+	$args = array(
+		'post_type'              => $post_type,
+		'post_status'            => 'publish',
+		'has_password'           => false,
+		'posts_per_page'         => 1,
+		'orderby'                => 'ID',
+		'order'                  => 'ASC',
+		'ignore_sticky_posts'    => true,
+		'update_post_term_cache' => false,
+		'update_post_meta_cache' => false,
```
Comment (westonruter) [inline src/wp-includes/query.php:1297]: More data is being queried than is needed:

```suggestion
		'update_post_meta_cache' => false,
		'fields'                 => 'ids',
```
